ipam: add metrics to track per node capacity

tommyp1ckles opened the pull request:
This PR adds metrics to help track IPAM Node capacity with more granularity.
Specifically, the goal is to help add alerting/dashboards for such questions such as:

* How many IPs are left on my node/cluster?
* How many/what Nodes have capacity?
* What is the capacity of any particular node?
 
This adds three new Operator metrics:

```
cilium_operator_ipam_available_ips
cilium_operator_ipam_used_ips
cilium_operator_ipam_needed_ips
```
As well, this seeks to deprecate the : `cilium_operator_ipam_ips` metric, aside from it being redundant it also is very confusing - the `type=available` does not track what you expect it to track.  

Which are all labelled by "target_node" (i.e. the name of the CiliumNode).

## How does this differ from the existing IPAM metrics?

There's a bit of a blind spot in the current set of metrics:
*  cilium_operator_ipam_ips: this metric is confusing, as well it should probably be separate metrics rather than a single one with classifying labels (CC: @chancez).
* empty_interface_slots/interface_candidates: These two metrics can answer whether there _is_ capacity left in the cluster but they aren't able to give any more detail about how much capacity is left, what nodes are full etc?
    * As well, they may be misleading.  For example: if you have 3 nodes each with 1 slot + 0 interface_candidates then its possible you might have: _3 possible ENIs X 10 Ips = 30 ... but if the nodes with the capacity have a different IP per interface limit then it could also be 10 x 50 = 500._   

```release-note
operator/ipam/metrics: Add new, more accurate, per-node available/used/needed metrics to deprecated existing ipam_ips metric.
```

tommyp1ckles committed:
ipam/metrics: add per node eni pam ip metrics.

These are intended to provide precise visibility into actual
available capacity individual nodes have for scheduling more CNI Endpoints.

Specifically this addresses problems with existing metrics:

* ipam_ips: do not actually provide total available for scheduling, but rather
	is the number of available ips for a particular point in time (discounting
	any future allocation).
* available interface / slots metrics: these can provide if there is capacity,
	but not specifically how much capacity.

We want to be able to determine exactly how many more CNI Pods should be able to
be scheduled, as well as what is the availability on individual nodes.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ipam: implement calculating node capacity for all cloud platforms.

While performing ResyncInterfacesAndIPs, calculate the theoretical
upper limit for number of IPAM usable addresses for each cloud platforms
IPAM implementation.

This includes:
* Known VM Instance type ENI/IPV4 limits.
* Primary interfaces/addresses.
* Prefix delegation/leftover prefix delegated addresses for AWS.

These are returned as a struct of InterfaceStats (among previous stats),
which is used in subsequent commits to be emitted as metrics.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/aws,azure,alibaba: decouple ipam eni actions from node impls.

All IPAM ENI node implementations have a reference to the ipam.Node
manager.

This makes the individual node implementations harder to test.
To fix this, this replaces *ipam.Node with the minimal interface
necessary for each implementation.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
operator/ipam: fix wording on function comment.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

gandro commented on pkg/alibabacloud/eni/node.go:
```diff
@@ -187,11 +187,14 @@ func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationA
+		return nil, stats, fmt.Errorf(errUnableToDetermineLimits)
 	}
+
+	stats.NodeCapacity = limits.IPv4 * (limits.Adapters - 1)
```
Could you leave a comment why we're subtracting one adapter? I'm not familiar with AlibabaCloud, but I assume it's because on AlibabaCloud we're not using the primary adapter for pod IPs (where as on AWS ENI, it can be used depending on the value of first-interface-index)?

tommyp1ckles replied:
Yeah, when doing IP allocation prep, the primary is [not considered](https://github.com/cilium/cilium/blob/7339b89e0db239facfc4a4777485f3c68ec414d0/pkg/alibabacloud/eni/node.go#L260-L262) so it gets excluded from the count.

I'll add a comment.

gandro commented on pkg/ipam/node_manager.go:
```diff
@@ -32,6 +32,17 @@ type CiliumNodeGetterUpdater interface {
+	// This does not account for currently used addresses.
+	NodeCapacity int
+
+	// RemainingAvailableInterfaceCount is the number of interfaces...todo
```
Leftover TODO comment

gandro commented on pkg/ipam/metrics/metrics.go:
```diff
@@ -21,7 +21,12 @@ type prometheusMetrics struct {
+	NeededIPs            *prometheus.GaugeVec
+	// Deprecated, will be removed in version 1.15.
+	// Use AvailableIPs, UsedIPs and NeededIPs instead.
+	IPsAllocated *prometheus.GaugeVec
```
Consider adding a note in the upgrade notes in documentation, I think we usually have a section for added and removed metrics.

tommyp1ckles replied:
Update upgrade doc, as well as metrics doc.

gandro reviewed (commented):
Awesome, thanks! I focused on the IPAM aspect, that change looks good to me. A few minor things that stood out to me.

christarazi reviewed (approved):
Nice work, clean code and easy to read commits!

I don't have any other comment besides a nit to follow and what Sebastian already pointed out. When reading some of the commits, I was initially confused by what "CNI pods" meant and then realized it meant "managed (by Cilium) pods". The latter is typically how we convey the relationship as you probably already know, but thought I'd call it out.

tommyp1ckles committed:
documentation: update metrics docs with ipam available/used/needed.

Updated metrics docs with new ipam metrics.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
upgrade: update upgrade guide with mention of new metrics.

Mention the addition of ipam_{available,used,needed} metrics.
As well, notify in upgrade guide about intended deprecation of ipam_ips metric.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

zacharysarah commented on Documentation/operations/upgrade.rst:
```diff
@@ -365,6 +368,7 @@ Deprecated Metrics
 * ``cilium_policy_import_errors_total`` is deprecated. Please use
   ``cilium_policy_change_total``, which counts all policy changes (Add, Update, Delete)
   based on outcome ("success" or "failure").
+* ``cilium_operator_ipam_ips`` is deprecated. Please use ``cilium_operator_ipam_{available,used,needed}_ips`` instead.
```
Use plain, direct language
```suggestion
* ``cilium_operator_ipam_ips`` is deprecated. Use ``cilium_operator_ipam_{available,used,needed}_ips`` instead.
```

zacharysarah commented on pkg/alibabacloud/eni/node.go:
```diff
@@ -183,11 +187,17 @@ func (n *Node) CreateInterface(ctx context.Context, allocation *ipam.AllocationA
+
+	// During preparation of IP allocations, the primary NIC is not considered
+	// for allocation.
+	// So we don't need to consider it for capacity calculation.
```
Nits for punctuation
```suggestion
	// During preparation of IP allocations the primary NIC is not considered
	// for allocation, so we don't need to consider it for capacity calculation.
```

zacharysarah commented on pkg/aws/eni/instances.go:
```diff
@@ -222,8 +222,9 @@ func (m *InstancesManager) UpdateENI(instanceID string, eni *eniTypes.ENI) {
 
-// ForeachInstance will iterate over each instance inside `instances`, and call
-// `fn`. This function is read-locked for the entire execution.
+// ForeachInstance will iterate over each interface for a particular instance inside `instances`,
```
Comma nit
```suggestion
// ForeachInstance will iterate over each interface for a particular instance inside `instances`
```

zacharysarah commented on pkg/aws/eni/node.go:
```diff
@@ -544,6 +556,19 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
 
+	// 1. This calculates the base interface effective limit on this Node, given:
+	// 		* IPAM Prefix Delegation.
+	// 		* Node Spec usePrimaryAddress being enabled.
```
Nit for periods on fragments
```suggestion
	// 		* IPAM Prefix Delegation
	// 		* Node Spec usePrimaryAddress being enabled
```

zacharysarah commented on pkg/aws/eni/node.go:
```diff
@@ -544,6 +556,19 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
+	//
+	_, stats.NodeCapacity = n.getEffectiveIPLimits(nil, limits.IPv4)
+
+	// 2. The base node limit is the number of adapters multiplied by the instances IP limit.
```
Appropriate period use. ✨

zacharysarah commented on pkg/ipam/types/stats.go:
```diff
@@ -0,0 +1,15 @@
+
+	// RemainingAvailableInterfaceCount is the number of interfaces currently available.
+	RemainingAvailableInterfaceCount int
+}
```
Well written. ✨

zacharysarah reviewed (commented):
@tommyp1ckles 👋🏻 This mostly LGTM from a docs perspective; my feedback amounts to being picky about tiny details.

gandro commented on pkg/ipam/types/stats.go:
```diff
@@ -0,0 +1,15 @@
+package types
```
Note that there are some linter errors in this file

tommyp1ckles replied:
fixed, thanks!

gandro reviewed (approved):
Thanks, looks good to me (besides the linter errors)!

tommyp1ckles committed:
ipam/aws: fix issue where aws/eni tests would panic due to nil iface

Previously IsPrefixDelegationEnabled implementation on *ipam.Node would check if the receiver node pointer was nil. If so it defaulted to false.

This is a bit dangerous as function on nil concrete types will still be invoked, whereas interface types will panic (i.e. because a nil interface doesn't actually have a function to lookup).

This removes that code, and defers nil checking to the caller.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

zacharysarah reviewed (approved):
@tommyp1ckles Thanks for the updates! ✨ LGTM
